src/System.Net.Http/src/System/Net/Http/Managed/AuthenticationHandler.cs

```diff
@@ -0,0 +1,89 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Net.Http.Headers;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Net.Http
+{
+    internal sealed class AuthenticationHandler : HttpMessageHandler
+    {
+        private readonly HttpMessageHandler _innerHandler;
+        private readonly bool _preAuthenticate;
+        private readonly ICredentials _credentials;
+
+        public AuthenticationHandler(bool preAuthenticate, ICredentials credentials, HttpMessageHandler innerHandler)
+        {
+            if (innerHandler == null)
+            {
+                throw new ArgumentNullException(nameof(innerHandler));
+            }
+
+            if (credentials == null)
+            {
+                throw new ArgumentNullException(nameof(credentials));
+            }
+
+            _preAuthenticate = preAuthenticate;
+            _credentials = credentials;
+            _innerHandler = innerHandler;
+        }
+
+        private bool TrySetBasicAuthToken(HttpRequestMessage request)
+        {
+            NetworkCredential credential = _credentials.GetCredential(request.RequestUri, "Basic");
+            if (credential == null)
+            {
+                return false;
+            }
+
+            request.Headers.Authorization = new AuthenticationHeaderValue("Basic", BasicAuthenticationHelper.GetBasicTokenForCredential(credential));
+            return true;
+        }
+
+        protected internal override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            if (_preAuthenticate)
+            {
+                TrySetBasicAuthToken(request);
+            }
+
+            HttpResponseMessage response = await _innerHandler.SendAsync(request, cancellationToken);
```

Comment:
`ConfigureAwait(false)` throughout. I'm sure you're already aware and this will be among the things addressed subsequently.